File: apps/flutter/lib/features/accounts/domain/entities/account.dart
class Account {
  final String id;
  final String name;
  final String type;
  final String? bankType;
  final String currency;
  final bool isActive;

  const Account({
    required this.id,
    required this.name,
    required this.type,
    this.bankType,
    required this.currency,
    required this.isActive,
  });

  factory Account.fromJson(Map<String, dynamic> json) {
    return Account(
      id: (json["accountId"] ?? json["id"]) as String,
      name: json["name"] as String,
      type: json["type"] as String,
      bankType: json["bankType"] as String?,
      currency: json["currency"] as String? ?? "BRL",
      isActive: json["isActive"] as bool? ?? true,
    );
  }
}

File: packages/database/src/models/account.ts
import { AggregateRoot } from "@abejarano/ts-mongodb-criteria";
import { createMongoId } from "./shared/mongo_id";

export enum AccountType {
  Cash = "cash",
  Bank = "bank",
  Wallet = "wallet",
  Broker = "broker",
  CreditCard = "credit_card",
}

export type AccountPrimitives = {
  id?: string;
  accountId: string;
  userId: string;
  name: string;
  type: AccountType;
  bankType?: string | null;
  currency: string;
  isActive: boolean;
  createdAt: Date;
  updatedAt?: Date;
};

export type AccountCreateProps = {
  userId: string;
  name: string;
  type: AccountType;
  bankType?: string | null;
  currency?: string;
  isActive?: boolean;
};

export class Account extends AggregateRoot {
  private readonly props: AccountPrimitives;

  private constructor(props: AccountPrimitives) {
    super();
    this.props = props;
  }

  static create(props: AccountCreateProps): Account {
    const now = new Date();

    return new Account({
      accountId: createMongoId(),
      userId: props.userId,
      name: props.name,
      type: props.type,
      bankType: props.bankType,
      currency: props.currency ?? "BRL",
      isActive: props.isActive ?? true,
      createdAt: now,
      updatedAt: now,
    });
  }

  static override fromPrimitives(primitives: AccountPrimitives): Account {
    return new Account(primitives);
  }

  getId(): string {
    return this.props.id ?? this.props.accountId;
  }

  getAccountId(): string {
    return this.props.accountId;
  }

  getBankType(): string | null {
    return this.props.bankType ?? null;
  }

  getType(): AccountType {
    return this.props.type;
  }

  toPrimitives(): AccountPrimitives {
    return this.props;
  }

  getCurrencry() {
    return this.props.currency;
  }
}
